### PR TITLE
AUTO-2018  Automation offchain lookup plugin

### PR DESCRIPTION
**Pull request by cmalec**

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+		}
+		r.lggr.Debugf("[OffchainLookup] cache miss UpkeepInfo: %+v\n", upkeepInfo)
+		r.upkeepCache.Set(upkeepId.String(), upkeepInfo, cache.DefaultExpiration)
+	}
+	return upkeepInfo, nil
+}
+
+// decodeOffchainLookup decodes the revert error ChainlinkAPIFetch(string url, bytes extraData, string[] jsonFields, bytes4 callbackSelector)
+func (r *EvmRegistry) decodeOffchainLookup(data []byte) (OffchainLookup, error) {
+	offchainLookup := OffchainLookup{}
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:147]: No big deal but normally I would drop this line, and then get rid of lines 155-158, and instead just instantiate the type's fields when returning, e.g.

```
return OffchainLookup{
		url: *abi.ConvertType(errorParameters[0], new(string)).(*string),
		extraData: *abi.ConvertType(errorParameters[1], new([]byte)).(*[]byte),
		fields: *abi.ConvertType(errorParameters[2], new([]string)).(*[]string),
		callbackFunction: *abi.ConvertType(errorParameters[3], new([4]byte)).(*[4]byte)
	}, nil
```

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+// offchainLookup looks through check upkeep results looking for any that need off chain lookup
+func (r *EvmRegistry) offchainLookup(ctx context.Context, upkeepResults []types.UpkeepResult) ([]types.UpkeepResult, error) {
+	for i, _ := range upkeepResults {
+		block, upkeepId, err := blockAndIdFromKey(upkeepResults[i].Key)
+		if err != nil {
+			r.lggr.Error("[OffchainLookup] error getting block and upkeep id:", err)
+			continue
+		}
+		// if its another reason continue/skip
+		if upkeepResults[i].FailureReason != UPKEEP_FAILURE_REASON_TARGET_CHECK_REVERTED {
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:50]: Could this `FailureReason` check go to the top of the loop, before calling `blockAndIdFromKey`?

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+		}
+
+		// checking if this upkeep is in cooldown from api errors
+		_, onIce := r.cooldownCache.Get(upkeepId.String())
+		if onIce {
+			r.lggr.Infof("[OffchainLookup] cooldown Skipping UpkeepId: %s\n", upkeepId)
+			continue
+		}
+
+		var offchainLookup OffchainLookup
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:61]: I think you might be able to drop this line, and then change the following line to

```
offchainLookup, err := r.decodeOffchainLookup(upkeepResults[i].PerformData)
```

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+		// if it doesn't decode to the offchain custom error continue/skip
+		offchainLookup, err = r.decodeOffchainLookup(upkeepResults[i].PerformData)
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:63]: For the following if blocks that check for an error, and set the FailureReason on the upkeep results, does it make sense to extract them into a single function that returns (performData, FailureReason, error)? Though you'd probably lose the more specific log messages as a consequence? 😬

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+	upkeepNeeded := *abi.ConvertType(unpack[0], new(bool)).(*bool)
+	if !upkeepNeeded {
+		return false, nil, nil
+	}
+	performData := *abi.ConvertType(unpack[1], new([]byte)).(*[]byte)
+	return true, performData, nil
+}
+
+func (r *EvmRegistry) doRequest(o OffchainLookup, upkeepId *big.Int, offchainConfig []byte) ([]byte, int, error) {
+	client := http.Client{}
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:224]: Should you specify a timeout on the client (I think by default there is no timeout)?

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+		return nil, 0, errors.Wrapf(err, "get request error:")
+	}
+
+	for _, key := range apiKeys.Keys {
+		value := key.DecryptVal
+		switch strings.ToLower(key.Type) {
+		case "header":
+			req.Header.Set(key.Name, value)
+		case "param":
+			newUrlString := strings.ReplaceAll(o.url, fmt.Sprintf("{%s}", key.Name), value)
```
Comment (ferglor) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:244]: Should the values be escaped? As in, `url.PathEscape(value)`?

**Comment by cmalec** [comment]
the registry changes will go away when 2.0.2 is merged

**Comment by cmalec** [comment]
> the registry changes will go away when 2.0.2 is merged

removed them to get tests passing

File: core/scripts/chaincli/handler/keeper_upkeep_events.go [hunk 7]
```diff
@@ -63,3 +64,53 @@ func (k *Keeper) UpkeepCounterEvents(ctx context.Context, hexAddr string, fromBl
...
+	upkeepCounter, err := upkeep_apifetch_wrapper.NewUpkeepAPIFetch(contractAddress, k.client)
+	if err != nil {
+		log.Fatalln("Failed to create a new upkeep pokemon", err)
+	}
+	filterOpts := bind.FilterOpts{
+		Start:   fromBlock,
+		End:     &toBlock,
+		Context: ctx,
+	}
+	upkeepIterator, err := upkeepCounter.FilterPokemonUpkeep(&filterOpts, nil)
```
Comment (FelixFan1992) [inline core/scripts/chaincli/handler/keeper_upkeep_events.go:79]: if the query takes too long, we can split the block range and batch them. but it seems filtering logs is very fast according to my Gauntlet experience.

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup_keys.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+type DecryptedValue struct {
+	UpkeepID *big.Int
+	Value    string
+}
+
+// getAPIKeys takes the offchain config and attempts to get any api keys that it may have
+func getAPIKeys(upkeepID *big.Int, offchainConfig []byte) (OffchainAPIKeys, error) {
+	offchainAPIKeys := OffchainAPIKeys{}
+	if len(offchainConfig) == 0 {
+		return OffchainAPIKeys{}, nil
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup_keys.go:37]: can just return `offchainAPIKeys`

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup_keys.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
+package evm
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/fxamacker/cbor/v2"
```
Comment (FelixFan1992) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup_keys.go:7]: just curious why cbor?

File: core/services/ocr2/plugins/ocr2keeper/evm/abi.go [hunk 10]
```diff
@@ -149,11 +172,156 @@ var (
...
 type performDataStruct struct {
 	CheckBlockNumber uint32   `abi:"checkBlockNumber"`
 	CheckBlockhash   [32]byte `abi:"checkBlockhash"`
 	PerformData      []byte   `abi:"performData"`
 }
+
+// KeeperRegistryInterface copied from wrapper
+//
+//go:generate mockery --quiet --name KeeperRegistryInterface --output ./mocks/ --case=underscore
+type KeeperRegistryInterface interface {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm/abi.go:187]: any reason to copy this out of the wrapper into this file?

File: core/services/ocr2/plugins/ocr2keeper/evm/registry.go [hunk 11]
```diff
@@ -497,6 +534,15 @@ func (r *EvmRegistry) doCheck(ctx context.Context, keys []types.UpkeepKey, chRes
 		return
 	}
 
+	// check for offchainLookup
+	upkeepResults, err = r.offchainLookup(ctx, upkeepResults)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm/registry.go:538]: let's put this behind a feature gate

File: core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+				marshal, marshalErr := gojq.Marshal(fieldValue)
+				if marshalErr != nil {
+					// if there is an issue marshalling default to setting the field value
+					results[i] = fmt.Sprint(fieldValue)
+					continue
+				}
+				results[i] = string(marshal)
+				continue
+			}
+			results[i] = fmt.Sprint(fieldValue)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm/offchain_lookup.go:349]: We should put some explicit limits on the size of fields allowed? Gets translated to gas used in callback

**Pull request closed without merging**